src/oc_erchef/apps/chef_objects/src/chef_cbv_cache.erl

```diff
@@ -0,0 +1,253 @@
+            % but we have been invoked anyway. `put` should only be invoked when the caller
+            % has first invoked `claim`.
+            {reply, {error, already_claimed}, State }
+    end.
+
+insert_into_cache(Tid, Key, Value, TTL) ->
+    case ets:insert_new(Tid, {Key, Value}) of
+        true ->
+            % - we see that often the pattern is that many requests for new keys
+            % come in clusters when traffic is first directed to the server.
+            % We can't avoid the initial cluster and corresponding CPU spike
+            % when we have nothing cached, but staggering the expirations after that
+            % will reduce CPU/VM utilization spikes that can impact the overall system.
+            erlang:send_after(TTL + rand:uniform(TTL), self(), {expire, Key});
+        false ->
+            % Value already exists.  Given the enforced ordering to prevent more than
+            % one caller from trying to put the same key, this should rarely occur
+            % (may happen if caller who invoked claim/1 then dies before put/2)
+            lager:info("chef_cbv_cache: Key ~p already present, ignoring.", [Key])
+    end.
+
+handle_info({expire, Key}, #state{tid = Tid} = State) ->
+    ets:delete(Tid, Key),
+    {noreply, State};
+handle_info({'EXIT', _From, Reason}, State) ->
+    lager:error("chef_cbv_cache: circuit breaker proc failed because ~p, restarting", [Reason]),
+    spawn_breaker(),
+    {noreply, State};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% Internal implementation
+
+spawn_breaker() ->
+    spawn_link(fun() -> breaker(0, false) end) ! init.
+
+breaker_tripped() ->
+  whereis(cbv_cache_breaker) ! { self(), check_limit },
+  receive
+    { result, limit_reached } ->
+      true;
+    { result, ok } ->
+      false
+  end.
+
+force_breaker(Disable) ->
+  whereis(cbv_cache_breaker) ! { force_disable, Disable }.
+
+% A circuit breaker that is spawn_linked to
+% cbv_cache, and monitors the state of the cache queue.
+% This prevents the overhead of checking the queue size on every request
+breaker(Count, Disabled) ->
+  receive
+    {Sender, check_limit} when Count >= ?MAX_QUEUE_LEN orelse Disabled =:= true ->
+      Sender ! {result, limit_reached},
+      breaker(Count, Disabled);
+    {Sender, check_limit} ->
+      Sender ! {result, ok},
+      breaker(Count, Disabled);
+    {force_disable, Disabled2} ->
+      breaker(Count, Disabled2);
+    init ->
+      register(cbv_cache_breaker, self()),
+      self() ! update,
+      breaker(Count, Disabled);
+    update ->
+      case whereis(chef_cbv_cache) of
+        undefined ->
+          Count2 = Count;
+        Pid ->
+          {_, Count2} = process_info(Pid, message_queue_len)
+      end,
+      erlang:send_after(?QUEUE_LEN_REFRESH_INTERVAL, self(), update),
```

Comment:
Would it be possible to enable spawn_breaker()/breaker() process only if 'cbv_cache_enabled = true' as 'erlang:send_after(?QUEUE_LEN_REFRESH_INTERVAL, self(), update)' makes this process busy unnecessarily if 'cbv_cache_enabled = false'.

Reply:
This can only be done if we do go ahead and move the config check to the caller process - otherwise the breaker needs to be running for `get` to check its status.